src/psyclone/psyir/transformations/matmul2code_trans.py

```diff
@@ -0,0 +1,286 @@
+            expected type.
+        :raises GenerationError: if the parent of the MATMUL \
+            operation is not an assignment.
+        :raises GenerationError: if the matmul arguments are not in \
+            the required form.
+
+        '''
+        # Check the supplied argument is a matvec node
+        if not isinstance(node, BinaryOperation):
+            raise GenerationError(
+                "The supplied node should be a BinaryOperation but "
+                "found '{0}'.".format(type(node).__name__))
+        if node.operator != BinaryOperation.Operator.MATMUL:
+            raise GenerationError(
+                "The supplied node should be a MATMUL BinaryOperation but "
+                "found '{0}'.".format(node.operator))
+        # Check the matmul is the only code on the rhs of an assignment
+        # i.e. ... = matmul(a,b)
+        if not isinstance(node.parent, Assignment):
+            raise GenerationError(
+                "Matmul2CodeTrans only supports the transformation of a "
+                "MATMUL operation when it is the sole operation on the rhs "
+                "of an assignment.")
+
+        matrix = node.children[0]
+        vector = node.children[1]
+        # The children of matvec should be References
+        if not (isinstance(matrix, Reference) and
+                isinstance(vector, Reference)):
+            raise GenerationError(
+                "Expected children of a MATMUL BinaryOperation to be "
+                "references, but found '{0}', '{1}'."
+                "".format(type(matrix).__name__,
+                          type(vector).__name__))
+
+        # The children of matvec should be References to arrays
+        if not (matrix.symbol.shape or vector.symbol.shape):
+            raise GenerationError(
+                "Expected children of a MATMUL BinaryOperation to be "
+                "references to arrays, but found '{0}', '{1}'."
+                "".format(type(matrix.symbol).__name__,
+                          type(vector.symbol).__name__))
+
+        # The first child (the matrix) should be declared as an array
+        # with at least 2 dimensions
+        if len(matrix.symbol.shape) < 2:
+            raise GenerationError(
+                "Expected 1st child of a MATMUL BinaryOperation to be "
+                "a matrix with at least 2 dimensions, but found '{0}'."
+                "".format(len(matrix.symbol.shape)))
+        if len(matrix.symbol.shape) == 2 and not matrix.children:
+            # If the matrix only has 2 dimensions and all of its data is
+            # used in the matrix vector multiply then the reference does
+            # not need to supply any dimension information.
+            pass
+        else:
+            # There should be one index per dimension. This is enforced
+            # by the array create method so is not tested here.
+
+            # The first two indices should be ranges. This
```

Comment:
Yes I kept it simple for the moment. I actually thought it would cause problems having other indices when I made this choice, but thinking about it, it should not matter which index is used from a correctness point of view.